elixir/apps/domain/lib/domain/repo.ex

```diff
@@ -156,15 +156,135 @@ defmodule Domain.Repo do
+
+  The function will be executed within the same transaction as the replace
+  operation and must return `{:ok, updated_schema}` or `{:error, reason}`.
+  """
+  @type on_replace :: (repo :: Ecto.Repo.t(),
+                       updated_schema :: Ecto.Schema.t(),
+                       replaced_schema :: Ecto.Schema.t() ->
+                         :ok)
+
+  @doc """
+  Uses query to fetch a single result from the database, locks it for update and
+  then either updates or replaces it using a changesets within a database transaction.
+  """
+  @spec fetch_and_update_or_replace(
+          queryable :: Ecto.Queryable.t(),
+          query_module :: module(),
+          opts ::
+            [
+              {:with, fetch_update_or_replace_changeset_fun()}
+              | {:preload, term()}
+              | {:filter, Domain.Repo.Filter.filters()}
+              | {:after_update_commit, update_after_commit() | [update_after_commit()]}
+              | {:after_replace_commit, replace_after_commit() | [replace_after_commit()]}
+              | {:on_replace, on_replace()}
+            ]
+            | Keyword.t()
+        ) ::
+          {:updated, Ecto.Schema.t()}
+          | {:replaced, Ecto.Schema.t(), Ecto.Schema.t()}
+          | {:error, :not_found}
+          | {:error, {:unknown_filter, metadata :: Keyword.t()}}
+          | {:error, {:invalid_type, metadata :: Keyword.t()}}
+          | {:error, {:invalid_value, metadata :: Keyword.t()}}
+          | {:error, Ecto.Changeset.t()}
+          | {:error, term()}
+  def fetch_and_update_or_replace(queryable, query_module, opts) do
+    {preload, opts} = Keyword.pop(opts, :preload, [])
+    {filter, opts} = Keyword.pop(opts, :filter, [])
+    {after_update_commit, opts} = Keyword.pop(opts, :after_update_commit, [])
+    {after_replace_commit, opts} = Keyword.pop(opts, :after_replace_commit, [])
+    {on_replace, opts} = Keyword.pop(opts, :on_replace, [])
+    {changeset_fun, transaction_opts} = Keyword.pop!(opts, :with)
+
+    with {:ok, queryable} <- Filter.filter(queryable, query_module, filter) do
+      Ecto.Multi.new()
+      |> Ecto.Multi.one(:fetch_and_lock, fn
+        _effects_so_far ->
+          Ecto.Query.lock(queryable, "FOR NO KEY UPDATE")
+      end)
+      |> Ecto.Multi.run(:changesets, fn
+        _repo, %{fetch_and_lock: schema} ->
+          {%Ecto.Changeset{} = update_changeset, replace_changeset_or_nil} =
+            changeset_fun.(schema)
+
+          {:ok, {update_changeset, replace_changeset_or_nil}}
+      end)
+      |> Ecto.Multi.update(:update, fn
+        %{changesets: {update_changeset, _replace_changeset_or_nil}} ->
+          update_changeset
+      end)
+      |> Ecto.Multi.run(:maybe_create, fn
+        _repo, %{changesets: {_update_changeset, nil}} ->
+          {:ok, nil}
+
+        repo, %{changesets: {_update_changeset, replace_changeset_or_nil}} ->
+          repo.insert(replace_changeset_or_nil)
+      end)
+      |> Ecto.Multi.run(:replace, fn
```

Comment:
This complexity comes from a need of not violating unique indexes while a record is replaced along with requirement of referencing the replacement record from a replaced one. The only way to do this is:
1. Update the record and mark is as deleted so that it's excluded from unique index;
2. Create a new one;
3. Update the "replaced by" ID on the deleted one.